### PR TITLE
Fix memory leak when using server-side

### PR DESCRIPTION
**Pull request by the83**
I recently ran into a memory leak when using this server-side. This occurs because the visibility handlers are set up in the constructor instead of a lifecycle hook, and because `componentWillUnmount` does not execute server-side, we never stop observing the element. I set up a simple example of this happening here: https://github.com/the83/memory-leak-example

This moves the setup into the `componentDidMount` hook so that we don't even observe visibility at all server-side.

### CODE REVIEW

File: src/createLoadableVisibilityComponent.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
-import React, { Component } from 'react'
+import * as React from 'react'
```
Comment (tazsingh) [inline src/createLoadableVisibilityComponent.js]: Would you be able to revert this change? I'd prefer to be explicit about imports.

File: src/__tests__/loadable-components/intersection-observer.test.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ const { IntersectionObserver, makeElementsVisible } = require('../../__mocks__/I
 global.IntersectionObserver = IntersectionObserver
 
 const loadableVisiblity = require('../../loadable-components')
+const trackedElements = require('../../tracked_elements').default
```
Comment (tazsingh) [inline src/__tests__/loadable-components/intersection-observer.test.js]: It looks like this is primarily being used for the tests. Would you be able to modify the `IntersectionObserver` mock to include this behaviour instead?

File: src/createLoadableVisibilityComponent.js [hunk 3]
```diff
@@ -95,7 +77,7 @@ function createLoadableVisibilityComponent (args, {
         return <div
           style={{display: 'inline-block', minHeight: '1px', minWidth: '1px'}}
           className={this.props.className}
-          ref={this.attachRef}
+          ref="loading"
```
Comment (tazsingh) [inline src/createLoadableVisibilityComponent.js]: Can you revert this change? String refs have been deprecated - https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs

File: src/createLoadableVisibilityComponent.js [hunk 4]
```diff
@@ -22,63 +22,45 @@ function createLoadableVisibilityComponent (args, {
   LoadingComponent,
 }) {
   let preloaded = false
-  const visibilityHandlers = []
```
Comment (tazsingh) [inline src/createLoadableVisibilityComponent.js]: Why did you remove this? It's to handle the case where `preload` is called after the component is part of the tree. In that scenario, we want all of the components to hit the underlying `LoadableComponent` to handle the interaction accordingly.

I believe by removing this and if `preload` is called after the component is mounted, then there will be a bug as it will fall back to the `IntersectionObserver` instead of already being loaded.

Of course, tests for this behaviour would be helpful 😅

**Review by tazsingh — CHANGES_REQUESTED** [review]
Good catch @the83! Thanks for the PR and apologies for the delayed response - I was on vacation for the past few weeks.

I've requested a few changes here. Let me know if you have any questions. But I'm happy to merge this and cut a release as soon as it's ready 👍

File: src/createLoadableVisibilityComponent.js (hunk 1 above)
Comment (the83) [reply src/createLoadableVisibilityComponent.js]: 👍

File: src/__tests__/loadable-components/intersection-observer.test.js (hunk 2 above)
Comment (the83) [reply src/__tests__/loadable-components/intersection-observer.test.js]: 👍

File: src/createLoadableVisibilityComponent.js (hunk 3 above)
Comment (the83) [reply src/createLoadableVisibilityComponent.js]: 👍

File: src/createLoadableVisibilityComponent.js (hunk 4 above)
Comment (the83) [reply src/createLoadableVisibilityComponent.js]: You're right, I think I was confused by the fact that the `this.state.visible` is set based on the `preloaded` var in the constructor, so I assumed that this was unnecessary.
 
I added the `visibilityHandlers` back in and [wrote a test](https://github.com/stratiformltd/react-loadable-visibility/pull/11/files#diff-f983afab730f1c0b922a76dceb0e5ca0R66) that demonstrates that the underlying component will be displayed when `preload` is called after mount.

File: src/__tests__/loadable-components/intersection-observer.test.js [hunk 5]
```diff
@@ -40,17 +45,63 @@ describe('Loadable', () => {
...
+    expect(trackedElements.length).toEqual(0)
+  })
+
   test('preload calls loadable load', () => {
     loadableVisiblity(opts).load()
 
     expect(loadable().load).toHaveBeenCalled()
   })
 
+  test('preload will cause the loadable component do be displayed', () => {
```
Comment (tazsingh) [inline src/__tests__/loadable-components/intersection-observer.test.js]: Thanks for adding this test! Can you add it for the `react-loadable` section as well?

Also minor typo: should be `to` instead of `do`

File: src/createLoadableVisibilityComponent.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ import React, { Component } from 'react'
 import { IntersectionObserver } from './capacities'
 
 let intersectionObserver
-let trackedElements = new Map()
+const trackedElements = new Map()
```
Comment (tazsingh) [inline src/createLoadableVisibilityComponent.js:5]: Can this now be removed as it's handled inside the mocked IntersectionObserver?

File: src/__mocks__/IntersectionObserver.js [hunk 7]
```diff
@@ -1,4 +1,7 @@
 const intersectionObservers = []
+const trackedElements = []
+
+module.exports.trackedElements = trackedElements
```
Comment (tazsingh) [inline src/__mocks__/IntersectionObserver.js]: I think the naming here is a bit confusing. Maybe `allTrackedElements` or `globallyTrackedElements` or similar to signify that it's a global that affects all instances of `IntersectionObserver`?

File: src/__mocks__/IntersectionObserver.js [hunk 8]
```diff
@@ -17,7 +20,7 @@ module.exports.IntersectionObserver = class IntersectionObserver {
   constructor(callback) {
     this.callback = callback
 
-    this.trackedElements = []
+    this.trackedElements = trackedElements
```
Comment (tazsingh) [inline src/__mocks__/IntersectionObserver.js]: I'm not a fan of initializing the `IntersectionObserver` with what was potentially previously tracked as it's not quite per spec. But I do like what you've done inside the `LoadableVisibilityComponent` itself where you double observed the tracked element via the `IntersectionObserver` and in `trackedElements`.

Perhaps a different approach would be something like:

```js
class IntersectionObserver {
  constructor(callback) {
    // ...
    this.trackedElements = []
    // ...
  }

  observe(element) {
    this.trackedElements.push(element)
    globallyTrackedElements.push(element)
  }

  unobserve(element) {
    // remove the element from this.trackedElements and globallyTrackedElements
  }
}
```

Then you can assert against `globallyTrackedElements` to ensure that there aren't any leaks 👍

**Review by tazsingh — CHANGES_REQUESTED** [review]
Almost there 🙂

File: src/createLoadableVisibilityComponent.js (hunk 6 above)
Comment (the83) [reply src/createLoadableVisibilityComponent.js:5]: I'm not sure I follow—what does the mocked IntersectionObserver have to do with the `trackedElements` in `createLoadableVisibilityComponent`?

Comment (tazsingh) [reply src/createLoadableVisibilityComponent.js:5]: Sorry I confused this with something else. Took another look and all is good 👍

**Review by tazsingh — APPROVED** [review]

**Comment by tazsingh** [comment]
Released under `2.5.0` - https://github.com/stratiformltd/react-loadable-visibility/releases/tag/v2.5.0
Thanks!
